File: src/main/java/net/gegy1000/terrarium/server/map/source/osm/OverpassSource.java
package net.gegy1000.terrarium.server.map.source.osm;

import net.gegy1000.terrarium.Terrarium;
import net.gegy1000.terrarium.server.map.osm.OsmDataParser;
import net.gegy1000.terrarium.server.map.source.CachedRemoteSource;
import net.gegy1000.terrarium.server.map.source.SourceException;
import net.gegy1000.terrarium.server.map.source.tiled.DataTilePos;
import net.gegy1000.terrarium.server.map.source.tiled.TiledSource;
import net.gegy1000.terrarium.server.util.Coordinate;
import net.gegy1000.terrarium.server.world.EarthGenerationSettings;
import org.apache.commons.io.IOUtils;
import org.apache.http.HttpEntity;
import org.apache.http.HttpRequestInterceptor;
import org.apache.http.HttpResponseInterceptor;
import org.apache.http.client.config.RequestConfig;
import org.apache.http.client.entity.GzipDecompressingEntity;
import org.apache.http.client.methods.CloseableHttpResponse;
import org.apache.http.client.methods.HttpPost;
import org.apache.http.entity.StringEntity;
import org.apache.http.impl.client.CloseableHttpClient;
import org.apache.http.impl.client.HttpClientBuilder;

import java.io.BufferedInputStream;
import java.io.BufferedReader;
import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.List;
import java.util.regex.Pattern;

public class OverpassSource extends TiledSource<OverpassTileAccess> implements CachedRemoteSource {
    private static final File CACHE_ROOT = new File(CachedRemoteSource.GLOBAL_CACHE_ROOT, "osm");

    private static final String OVERPASS_ENDPOINT = "http://www.overpass-api.de/api/interpreter";

    private final CloseableHttpClient client = HttpClientBuilder.create()
            .addInterceptorFirst((HttpRequestInterceptor) (request, context) -> {
                request.setHeader("Accent-Encoding", "gzip");
                request.setHeader("User-Agent", Terrarium.MODID);
                request.setHeader("Referer", "https://github.com/gegy1000/Terrarium");
            }).addInterceptorFirst((HttpResponseInterceptor) (response, context) -> {
                HttpEntity entity = response.getEntity();
                Arrays.stream(entity.getContentEncoding().getElements())
                        .filter(element -> element.getName().equalsIgnoreCase("gzip"))
                        .findFirst()
                        .ifPresent(element -> response.setEntity(new GzipDecompressingEntity(entity)));
            })
            .setDefaultRequestConfig(RequestConfig.custom()
                    .setConnectTimeout(1000)
                    .setConnectionRequestTimeout(1000)
                    .setSocketTimeout(30000)
                    .build())
            .build();

    private final EarthGenerationSettings settings;

    private final File cacheRoot;

    private final String queryLocation;
    private final int queryVersion;

    private final boolean shouldSample;

    private String query;

    public OverpassSource(EarthGenerationSettings settings, double tileSize, String cacheRoot, String queryLocation, int queryVersion) {
        super(tileSize, 4);
        this.settings = settings;
        this.cacheRoot = new File(CACHE_ROOT, cacheRoot);
        this.queryLocation = queryLocation;
        this.queryVersion = queryVersion;

        this.shouldSample = Coordinate.fromLatLng(settings, tileSize, tileSize).getBlockX() > 512;
    }

    public void loadQuery() {
        String queryLocation = this.queryLocation;
        try (BufferedReader input = new BufferedReader(new InputStreamReader(OverpassSource.class.getResourceAsStream(queryLocation)))) {
            StringBuilder query = new StringBuilder();
            List<String> lines = IOUtils.readLines(input);
            for (String line : lines) {
                query.append(line);
            }
            this.query = query.toString();
        } catch (IOException e) {
            Terrarium.LOGGER.error("Failed to load Overpass query at {}", queryLocation, e);
        }
    }

    @Override
    public File getCacheRoot() {
        return this.cacheRoot;
    }

    @Override
    public InputStream getRemoteStream(DataTilePos key) throws IOException {
        HttpPost post = new HttpPost(OVERPASS_ENDPOINT);
        double minLatitude = this.getLatitude(key) - 0.0005;
        double minLongitude = this.getLongitude(key) - 0.0005;
        double maxLatitude = this.getMaxLatitude(key) + 0.0005;
        double maxLongitude = this.getMaxLongitude(key) + 0.0005;

        String bbox = String.format("%.6f,%.6f,%.6f,%.6f", minLatitude, minLongitude, maxLatitude, maxLongitude);
        String formattedQuery = this.query.replaceAll(Pattern.quote("{{bbox}}"), bbox);
        post.setEntity(new StringEntity(formattedQuery));

        CloseableHttpResponse response = this.client.execute(post);
        if (response.getStatusLine().getStatusCode() == 429) {
            try {
                // TODO: Handle rate limit better
                response.close();
                Thread.sleep(150);
                return this.getRemoteStream(key);
            } catch (InterruptedException e) {
                Terrarium.LOGGER.error("Interrupted while awaiting rate limit", e);
            }
        }

        return response.getEntity().getContent();
    }

    @Override
    public String getCachedName(DataTilePos key) {
        return String.format("%s_%s.osm", key.getTileX(), key.getTileY());
    }

    @Override
    public OverpassTileAccess loadTile(DataTilePos key) throws SourceException {
        return this.loadTile(key, 0);
    }

    private OverpassTileAccess loadTile(DataTilePos key, int retries) throws SourceException {
        try {
            return OsmDataParser.parse(this.getStream(key));
        } catch (IOException e) {
            Terrarium.LOGGER.error("Failed to load overpass map tile at {}", this.getCachedName(key), e);
        } catch (RuntimeException e) {
            Terrarium.LOGGER.error("Failed to parse overpass map tile at {}, reloading", this.getCachedName(key), e);
            this.removeCache(key);
            if (retries < 2) {
                return this.loadTile(key, retries + 1);
            }
        }
        return null;
    }

    @Override
    protected OverpassTileAccess getDefaultTile() {
        return new OverpassTileAccess();
    }

    @Override
    public EarthGenerationSettings getSettings() {
        return this.settings;
    }

    @Override
    public void cacheMetadata(DataTilePos key) {
        File metadataFile = new File(this.getCacheRoot(), String.format("%s_%s.meta", key.getTileX(), key.getTileY()));
        try (DataOutputStream output = new DataOutputStream(new FileOutputStream(metadataFile))) {
            output.writeShort(this.queryVersion);
        } catch (IOException e) {
            Terrarium.LOGGER.error("Failed to cache OSM tile metadata at {}", key, e);
        }
    }

    @Override
    public boolean shouldLoadCache(DataTilePos key, File file) {
        if (file.exists()) {
            File metadataFile = new File(this.getCacheRoot(), String.format("%s_%s.meta", key.getTileX(), key.getTileY()));
            if (metadataFile.exists()) {
                try (DataInputStream input = new DataInputStream(new BufferedInputStream(new FileInputStream(metadataFile)))) {
                    return input.readUnsignedShort() == this.queryVersion;
                } catch (IOException e) {
                    Terrarium.LOGGER.error("Failed to write OSM tile metadata at {}", key, e);
                }
            }
        }
        return false;
    }

    public boolean shouldSample() {
        return this.shouldSample;
    }

    private double getLatitude(DataTilePos pos) {
        return -pos.getTileY() * this.getTileSize();
    }

    private double getLongitude(DataTilePos pos) {
        return pos.getTileX() * this.getTileSize();
    }

    private double getMaxLatitude(DataTilePos pos) {
        return this.getLatitude(pos) + this.getTileSize();
    }

    private double getMaxLongitude(DataTilePos pos) {
        return this.getLongitude(pos) + this.getTileSize();
    }
}

File: src/main/kotlin/net/gegy1000/terrarium/server/map/source/osm/OverpassSource.kt
package net.gegy1000.terrarium.server.map.source.osm

import com.google.gson.Gson
import com.google.gson.JsonParser
import net.gegy1000.terrarium.Terrarium
import net.gegy1000.terrarium.server.map.source.CachedRemoteSource
import net.gegy1000.terrarium.server.map.source.tiled.DataTilePos
import net.gegy1000.terrarium.server.map.source.tiled.TiledSource
import net.gegy1000.terrarium.server.util.Coordinate
import net.gegy1000.terrarium.server.world.EarthGenerationSettings
import net.minecraft.util.math.MathHelper
import org.apache.http.HttpRequest
import org.apache.http.HttpResponse
import org.apache.http.client.entity.GzipDecompressingEntity
import org.apache.http.client.methods.HttpPost
import org.apache.http.entity.StringEntity
import org.apache.http.impl.client.CloseableHttpClient
import org.apache.http.impl.client.HttpClientBuilder
import java.io.DataInputStream
import java.io.DataOutputStream
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.io.IOException
import java.io.InputStream
import java.io.InputStreamReader

class OverpassSource(override val settings: EarthGenerationSettings) : TiledSource<OverpassTileAccess>(SIZE_DEGREES, 4), CachedRemoteSource {
    companion object {
        const val SIZE_DEGREES = 0.2

        const val OVERPASS_ENDPOINT = "http://www.overpass-api.de/api/interpreter"
        const val QUERY_LOCATION = "/assets/terrarium/query/overpass_query.oql"

        const val QUERY_VERSION = 1

        val GSON = Gson()
        val JSON_PARSER = JsonParser()

        val client: CloseableHttpClient = HttpClientBuilder.create()
                .addInterceptorFirst { request: HttpRequest, _ ->
                    if (!request.containsHeader("Accept-Encoding")) {
                        request.addHeader("Accept-Encoding", "gzip")
                    }
                    if (!request.containsHeader("User-Agent")) {
                        request.addHeader("User-Agent", Terrarium.MODID)
                    }
                }
                .addInterceptorFirst { response: HttpResponse, _ ->
                    val entity = response.entity
                    entity.contentEncoding.elements.forEach {
                        if (it.name.equals("gzip", ignoreCase = true)) {
                            response.entity = GzipDecompressingEntity(entity)
                            return@forEach
                        }
                    }
                }
                .build()
    }

    private val DataTilePos.latitude: Double
        get() = -tileY * OverpassSource.SIZE_DEGREES
    private val DataTilePos.longitude: Double
        get() = tileX * OverpassSource.SIZE_DEGREES

    private val DataTilePos.maxLatitude: Double
        get() = latitude + OverpassSource.SIZE_DEGREES
    private val DataTilePos.maxLongitude: Double
        get() = longitude + OverpassSource.SIZE_DEGREES

    lateinit var query: String

    override val defaultTile
        get() = OverpassTileAccess()
    override val cacheRoot = File(CachedRemoteSource.globalCacheRoot, "osm")

    fun loadQuery() {
        val input = InputStreamReader(OverpassSource::class.java.getResourceAsStream(QUERY_LOCATION))
        query = input.use { it.readText().replace("\n", "") }
    }

    fun getTilePos(coordinate: Coordinate): DataTilePos {
        val tileX = MathHelper.floor(coordinate.longitude / OverpassSource.SIZE_DEGREES)
        val tileZ = MathHelper.ceil(-coordinate.latitude / OverpassSource.SIZE_DEGREES)
        return DataTilePos(tileX, tileZ)
    }

    fun sampleArea(minCoordinate: Coordinate, maxCoordinate: Coordinate): OverpassTileAccess {
        val minTilePos = getTilePos(minCoordinate)
        val maxTilePos = getTilePos(maxCoordinate)

        val elements = hashSetOf<Element>()

        for (tileZ in minTilePos.tileY..maxTilePos.tileY) {
            for (tileX in minTilePos.tileX..maxTilePos.tileX) {
                elements.addAll(getTile(DataTilePos(tileX, tileZ)).elements)
            }
        }

        return OverpassTileAccess(elements)
    }

    override fun loadTile(pos: DataTilePos): OverpassTileAccess? {
        val input = getStream(pos)

        try {
            val elements = HashSet<Element>()

            val root = JSON_PARSER.parse(InputStreamReader(input)).asJsonObject
            val elementsArray = root["elements"].asJsonArray

            elementsArray.asSequence().map { it.asJsonObject }.forEach {
                val tags = HashMap<String, String>()
                val nodes = ArrayList<Int>()

                val id = it["id"].asInt
                val type = it["type"].asString
                val latitude = it["lat"]?.asDouble ?: 0.0
                val longitude = it["lon"]?.asDouble ?: 0.0

                if (it.has("tags")) {
                    val tagsObject = it["tags"].asJsonObject
                    tagsObject.entrySet().asSequence().filter { it.value.isJsonPrimitive }.forEach {
                        tags.put(it.key, it.value.asString)
                    }
                }

                if (it.has("nodes")) {
                    nodes.addAll(it["nodes"].asJsonArray.map { it.asInt })
                }

                elements.add(Element(id, type, latitude, longitude, nodes, tags))
            }

            return OverpassTileAccess(elements)
        } catch (e: Exception) {
            Terrarium.LOGGER.error("Failed to download overpass map tile ${getCachedName(pos)}", e)
        } finally {
            input.close()
        }

        return null
    }

    override fun getRemoteStream(key: DataTilePos): InputStream {
        val post = HttpPost(OVERPASS_ENDPOINT)
        post.entity = StringEntity(query.format(key.latitude, key.longitude, key.maxLatitude, key.maxLongitude))

        val response = client.execute(post)
        if (response.statusLine.statusCode == 429) {
            // TODO: Handle rate limit better
            response.close()
            Thread.sleep(150)
            return getRemoteStream(key)
        }

        return response.entity.content
    }

    override fun getCachedName(key: DataTilePos) = "${key.tileX}_${key.tileY}.osm"

    override fun cacheMetadata(key: DataTilePos) {
        val metadataFile = File(cacheRoot, "${key.tileX}_${key.tileY}.meta")
        val metadataOutput = DataOutputStream(FileOutputStream(metadataFile))

        try {
            metadataOutput.writeShort(QUERY_VERSION)
        } catch (e: IOException) {
            Terrarium.LOGGER.error("Failed to cache OSM tile metadata at $key")
        } finally {
            metadataOutput.close()
        }
    }

    override fun shouldLoadCache(key: DataTilePos, file: File): Boolean {
        if (file.exists()) {
            val metadataFile = File(cacheRoot, "${key.tileX}_${key.tileY}.osm.meta")
            if (metadataFile.exists()) {
                val metadataInput = DataInputStream(FileInputStream(metadataFile).buffered())
                val version = metadataInput.use { metadataInput.readUnsignedShort() }
                return version == QUERY_VERSION
            }
        }
        return false
    }

    data class Element(val id: Int, val type: String, val latitude: Double, val longitude: Double, val nodes: List<Int>, val tags: Map<String, String>) {
        fun isType(key: String, value: String) = this.tags[key] == value

        override fun equals(other: Any?) = other is Element && other.id == id

        override fun hashCode() = id
    }
}
